DigitalLearningSolutions.Data.Tests/Services/GroupServiceTests/GroupsServiceAddCourseTests.cs

```diff
@@ -0,0 +1,597 @@
+            const int adminId = 1;
+            const int completeWithinMonths = 3;
+            var groupCourse = GroupTestHelper.GetDefaultGroupCourse(
+                completeWithinMonths: completeWithinMonths,
+                supervisorAdminId: adminId
+            );
+            var expectedFutureDate = testDate.AddMonths(3);
+            SetupEnrolProcessFakes(
+                GenericNewProgressId,
+                GenericRelatedTutorialId
+            );
+            SetUpAddCourseEnrolProcessFakes(groupCourse);
+
+            // When
+            groupsService.AddCourseToGroup(
+                groupCourse.GroupId,
+                groupCourse.CustomisationId,
+                8,
+                adminId,
+                true,
+                adminId
+            );
+
+            // Then
+            using (new AssertionScope())
+            {
+                DelegateProgressRecordMustNotHaveBeenUpdated();
+                A.CallTo(
+                    () => progressDataService.CreateNewDelegateProgress(
+                        reusableGroupDelegate.DelegateId,
+                        reusableGroupCourse.CustomisationId,
+                        reusableGroupCourse.CurrentVersion,
+                        testDate,
+                        3,
+                        adminId,
+                        expectedFutureDate,
+                        adminId
+                    )
+                ).MustHaveHappened();
+                A.CallTo(() => progressDataService.CreateNewAspProgress(GenericRelatedTutorialId, GenericNewProgressId))
+                    .MustHaveHappened();
+            }
+        }
+
+        [Test]
+        public void
+            AddCourseToGroup_keeps_existing_supervisor_when_updating_progress_when_course_supervisor_id_is_not_null()
+        {
+            // Given
+            const int adminId = 1;
+            const int supervisorId = 12;
+            var groupCourse = GroupTestHelper.GetDefaultGroupCourse(supervisorAdminId: supervisorId);
+            SetupEnrolProcessFakes(
+                GenericNewProgressId,
+                GenericRelatedTutorialId,
+                reusableProgressRecord
+            );
+            SetUpAddCourseEnrolProcessFakes(groupCourse);
+
+            // When
+            groupsService.AddCourseToGroup(
+                reusableGroupCourse.GroupId,
+                reusableGroupCourse.CustomisationId,
+                8,
+                adminId,
+                true,
+                supervisorId
+            );
+
+            // Then
+            using (new AssertionScope())
+            {
+                NewDelegateProgressRecordMustNotHaveBeenAdded();
+                A.CallTo(
+                    () => progressDataService.UpdateProgressSupervisorAndCompleteByDate(
+                        reusableProgressRecord.ProgressId,
+                        adminId,
+                        A<DateTime?>._
+                    )
+                ).MustHaveHappened();

```

Comment:
Is this actually testing that the supervisor id remains the same? Given the title, I would have expected the other supervisor Id to have been present in the call, but then it would seem the new and old supervisor id would be the same?

Reply:
Yes, we're passing in supervisorId 12 into the method, the group course has that supervisorId, but we expect the call to have the supervisor parameter = progressRecord.SupervisorId = adminId = 1 rather than 12.

Reply:
Okay, then yes this looks correct. I think I've gotten confused with the use of the adminId const here. 

I think we can make this test slightly more secure by changing the parameters around a bit. Looking at this from the test's perspective of "I don't know the details what's going on, but I am checking the expected results (calls)" then you could conclude that the progress records supervisor id is replaced with the adminId from the addedToAdminId parameter in the AddCourseToGroup call.

- Change the value of the addedToAdminId parameter to be different to the progress records' supervisor id
- Change the expected value of the supervisor id in UpdateProgressSupervisorAndCompleteByDate to be reusableProgressRecord.SupervisorId, just so that it is obvious that it came from the record.

Is that sensible?